### ISSUE TITLE
Variable Reduction [BUG] 

### ISSUE THREAD

**Issue by lfraser1** [ApexRMS/wisdm#21]
**Describe the bug**
Variable reduction does not run and puts out "Error in round(devExp, 2) : non-numeric argument to mathematical function
Execution halted." No variable reduction screen appears. It seems like the issue may be produced by the site data preparation step (see screenshot of datasheet). The site data appears to have multiple values for a variable for an individual site, or values do not match input rasters.   

**To Reproduce**
I am sending the library that produced this error to Katie. 

**Expected behavior**
I expect site preparation to pull values of each variable at each site. Then I expect the variable reduction screen to pop up showing a matrix of correlations between site variables for my points. 

**Screenshots**
![image](https://github.com/user-attachments/assets/61db827d-5f49-4f3b-a18d-2b7e78b2fe25)

**Relevant log output**
SyncroSim Version: 3.0.2.0
Operating System: Microsoft Windows NT 6.2.9200.0

Packages:
core -> 3.0.2
wisdm -> 2.0.1

Parent Scenario is: [10] Brassica tournifortii - steps 3-6
Result scenario is: [30] Brassica tournifortii - steps 3-6 ([10] @ 04-Oct-2024 9:28 AM)

--------------------------------------------
STARTING SIMULATION: 10/4/2024 : 9:28:52 AM
--------------------------------------------

Conda installation: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda
Conda environment: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda\envs\wisdm-1\wisdm-conda-s3

Only one field data observation present per pixel; no aggregation or weighting required.

3 - Site Data Preparation => Total time: 00:02:17

Conda installation: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda
Conda environment: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda\envs\wisdm-1\wisdm-conda-s3

4 - Background Data Generation => Total time: 00:00:30

Conda installation: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda
Conda environment: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda\envs\wisdm-1\wisdm-conda-s3

5 - Prepare Training/Testing Data => Total time: 00:00:24

Conda installation: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda
Conda environment: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda\envs\wisdm-1\wisdm-conda-s3

6 - Variable Reduction => Total time: 00:00:27
[Failure] The external program failed.
Program name was: C:\Users\lfraser\AppData\Local\SyncroSim3\Conda\envs\wisdm-1\wisdm-conda-s3\Scripts\Rscript.exe
Arguments were: "C:\Users\lfraser\AppData\Local\SyncroSim3\Packages\wisdm\2.0.1\6-variable-reduction.R"
Exit code was: 1
More information:
Attaching package: 'dplyr'
The following objects are masked from 'package:stats':
    filter, lag
The following objects are masked from 'package:base':
    intersect, setdiff, setequal, union
Saved
Error in round(devExp, 2) : non-numeric argument to mathematical function
Execution halted

--------------------------------------------
SIMULATION COMPLETE: 10/4/2024 : 9:32:31 AM
--------------------------------------------
Total simulation time: 00:03:39
**Software versions:**
 - SyncroSim: 3.0.2
 - WISDM 2.0.1

**Comment by katieb1** [comment]
@lfraser1 from the screenshot it looks like there is a single row per site/variable, which I think is what we would expect. One thing that seems weird to me are the extremely large negative numbers for some variables. I think these numbers are probably NO DATA, but are getting interpreted as numeric values by SyncroSim. Unfortunately the library you shared with me is too big for me to download. Is there a smaller library that you can repro the problem? Or can you create the backup without including the external outputs?

**Comment by katieb1** [comment]
Looking at the gHM raster in QGIS, I can see that the NO DATA value is indeed this really big negative number. 
![image](https://github.com/user-attachments/assets/09b7a890-0373-49a0-8559-4694ee81ff32)

When I look at a covariate raster from the WISDM example library, the NO DATA value is -9999. I wonder if WISDM is expecting NO DATA to be -9999 and that's what's causing the issue.
![image](https://github.com/user-attachments/assets/692b58f0-19fc-45d0-a04e-7accd6d0844e)

**Comment by lfraser1** [comment]
Yes, let me see if Catherine can help me delete superfluous runs and then repair so there are fewer versions of the rasters.  I think I put in the raw rasters in the folder that I shared before, so I'll try to share the backup without the external output.


Linnea Fraser (she/her)
SCC - Biological Science Technician

U.S. Geological Survey
Fort Collins Science Center
2150 Center Ave Bldg C
________________________________
From: Katie Birchard ***@***.***>
Sent: Wednesday, October 23, 2024 10:45 AM
To: ApexRMS/wisdm ***@***.***>
Cc: Fraser, Linnea (Contractor) ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [ApexRMS/wisdm] Variable Reduction [BUG] (Issue #21)




 This email has been received from outside of DOI - Use caution before clicking on links, opening attachments, or responding.



@lfraser1<https://github.com/lfraser1> from the screenshot it looks like there is a single row per site/variable, which I think is what we would expect. One thing that seems weird to me are the extremely large negative numbers for some variables. I think these numbers are probably NO DATA, but are getting interpreted as numeric values by SyncroSim. Unfortunately the library you shared with me is too big for me to download. Is there a smaller library that you can repro the problem? Or can you create the backup without including the external outputs?

**Comment by lfraser1** [comment]
Hmm I can navigate to Library>SyncroSim Core> and then backup, but the options are greyed out. They won't let me unselect Include external files in backup.


Linnea Fraser (she/her)
SCC - Biological Science Technician

U.S. Geological Survey
Fort Collins Science Center
2150 Center Ave Bldg C
________________________________
From: Katie Birchard ***@***.***>
Sent: Wednesday, October 23, 2024 10:45 AM
To: ApexRMS/wisdm ***@***.***>
Cc: Fraser, Linnea (Contractor) ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [ApexRMS/wisdm] Variable Reduction [BUG] (Issue #21)




 This email has been received from outside of DOI - Use caution before clicking on links, opening attachments, or responding.



@lfraser1<https://github.com/lfraser1> from the screenshot it looks like there is a single row per site/variable, which I think is what we would expect. One thing that seems weird to me are the extremely large negative numbers for some variables. I think these numbers are probably NO DATA, but are getting interpreted as numeric values by SyncroSim. Unfortunately the library you shared with me is too big for me to download. Is there a smaller library that you can repro the problem? Or can you create the backup without including the external outputs?

**Comment by katieb1** [comment]
Ok, not sure why the option is greyed out, but I just sent you a new release to test over email that might fix this bug. Basically it allows for NO DATA values in the covariate rasters other than -9999. I think before investing time trying to download and run this library, it's worth you testing the potential fix on your end first. Let me know if this still doesn't fix things!

**Comment by afilazzola** [comment]
This one's fixed.

The literal error from the report — Error in round(devExp, 2) : non-numeric argument to mathematical function — came from devExp containing an R try-error object (when the deviance-explained calculation failed for a covariate, e.g. due to degenerate/duplicate site values) and then trying to round() it. The fix adds exactly one line before the round() call in src/6-variable-reduction.R:253:

devExp <- sapply(devExp, function(x) if (inherits(x, "try-error")) NA else x)
devExp <- round(devExp, 2)

So any failed per-covariate GAM fit now becomes NA instead of crashing the whole variable-reduction stage.
